Auto-scaling of standalone Spark cluster (resolves #152)

hannes-ucsc opened the pull request:
Resolves #152 

We now distinguish between *dynamic* ("Spark-on-Toil") and *static* clusters. Within the latter we further distinguish between *standalone* and *external*. A standalone cluster is managed by the überscript. This PR obsoletes the `--add-to-etc-hosts` parameter and replaces it with `--master_ip=auto` to indicate standalone clusters. That special string is resolved in the Toil scripts using SimpleDB (the SparkMasterAddress model) and injected into the driver containers via Docker's `--add-host` option. For that I had to dig a little deeper into my OO magic bag of trick. Apologies for that.

This PR creates the standalone Spark master when it is needed. This has the advantage that it will be recreated if needed and that it does not sit idle during alignment.

This has not been tested. The doctests pass, though, needless to say.

hannes-ucsc commented on src/toil_scripts/adam_uberscript/automated_scaling.py:
```diff
@@ -149,6 +150,161 @@ def change_size(cls, conn, dom, n):
+
+    def release(self, delta=1):
+        """
+        Repeatedly attempt to decrease the semaphore by the given amount until successful. 
```
"increase"

hannes-ucsc commented:
@fnothaft, could you take a peek?

fnothaft commented on src/toil_scripts/adam_pipeline/adam_preprocessing.py:
```diff
@@ -24,37 +24,75 @@
+        real one, if they are different.
+        """
+        if self != self.actual:
+            add_host_option = '--add-host=spark-master:' + self.actual
```
Ah, savvy! Good call.

fnothaft commented on src/toil_scripts/adam_pipeline/adam_preprocessing.py:
```diff
@@ -81,36 +122,40 @@ def call_adam(masterIP, inputs, arguments):
+    masterIP = masterIP.actual
     if sparkOnToil:
-        containerID = check_output(["ssh", "-o", "StrictHostKeyChecking=no", masterIP, "docker", "ps", \
+        containerID = check_output(["ssh", "-o", "StrictHostKeyChecking=no", masterIP, "docker", "ps",
```
Small nit: if I'm splitting a call/array across multiple lines, I'll typically break it out into one line per element/logical group of elements. E.g., this would become something like:

```
containerID = check_output(["ssh",
"-o", "StrictHostKeyChecking=no",
masterIP,
"docker", "ps", "|",
"grep", "apache-hadoop-master", "|",
"awk", "'{print $1}'"])
```

Same for the line below.

hannes-ucsc replied:
Agreed. The grep and awk pipe should be replaced with pure Python, too.

fnothaft commented on src/toil_scripts/adam_uberscript/adam_uberscript.py:
```diff
@@ -220,110 +218,179 @@ def get_cluster_size(cluster_name):
+                num_workers = len(list_nodes(params.cluster_name, role='spark-slave'))
+                num_masters = len(list_nodes(params.cluster_name, role='spark-master'))
+                if num_masters == 0:
+                    # Absence of master indicates absence of cluster, so create it now.
```
Ah, nice! So, this will _auto start_ the Spark cluster as well?

hannes-ucsc replied:
Yes.

fnothaft commented on src/toil_scripts/adam_uberscript/adam_uberscript.py:
```diff
@@ -220,110 +218,179 @@ def get_cluster_size(cluster_name):
+                    output = check_output(['cgcloud',
+                                           'create-cluster',
+                                           '--list',
+                                           '--zone', aws_region + 'a',
```
Do we want to hard code this?

hannes-ucsc replied:
No. I'll add a parameter for this.

fnothaft commented on src/toil_scripts/adam_uberscript/adam_uberscript.py:
```diff
@@ -220,110 +218,179 @@ def get_cluster_size(cluster_name):
+                                           '--cluster-name', params.cluster_name,
+                                           '--leader-instance-type', params.spark_master_type or params.leader_type,
+                                           '--instance-type', params.spark_instance_type or params.instance_type,
+                                           '--num-workers', str(num_desired_workers),  # TODO: spot options
```
Make a ticket to track for later?

hannes-ucsc replied:
https://github.com/BD2KGenomics/toil-scripts/issues/174

fnothaft commented on src/toil_scripts/adam_uberscript/adam_uberscript.py:
```diff
@@ -220,110 +218,179 @@ def get_cluster_size(cluster_name):
+                                     cluster_name=params.cluster_name,
+                                     cluster_type='spark')
+                    elif num_desired_workers < num_workers:
+                        pass  # TODO: think about scaling down
```
Can't we reuse the toil cluster shrinking code?

hannes-ucsc replied:
To remove instances, sure. But what about the HDFS contents? Don't we need to to some rebalancing before taking instances out of service?

fnothaft replied:
Ah, fair... That gets a bit tricky...

fnothaft commented:
Generally LGTM, other than a few nits. Do we want to merge this before we have autoshrink implemented?

hannes-ucsc commented:
@fnothaft, I'd prefer to merge ASAP (after I make the changes you proposed, except auto-shrinking) and ask Audrey to test this today or tomorrow at the latest.

hannes-ucsc commented:
Addressed all comments. Pinging @fnothaft for re-review with a focus on `static_adam_preprocessing_dag()` and `remove_file()` to make sure I didn't hose those when resolving conflicts.

fnothaft commented:
@hannes-ucsc on a call, will review in ~30m.

fnothaft commented:
LGTM!

hannes-ucsc commented:
One additional last-minute change in the last commit:

https://github.com/BD2KGenomics/toil-scripts/commit/1bf16122d4c42528d135a4d11a6cdb7b4cda813f#diff-7060afbea50ff85e89fe57836bc27314R564

It makes --zone optional if CGCLOUD_ZONE is defined.

fnothaft commented:
As an aside @hannes-ucsc, when you said that this was untested in the PR body, I assumed you meant _external Spark cluster autoscaling_, not the modifications to the existing cluster autoscaling code. https://github.com/BD2KGenomics/toil-scripts/commit/df778f8c5e43edf0d22e479188a5aa18e17b8632 contains errors that break the core autoscaling loop. I'm fixing these in #181. That being said, I wouldn't have given an "OK to merge" vote if I had known that the existing autoscaling loop wasn't testet. This is my fault for not asking a question to clarify that, but if code is untested, it should be tagged with `needs work`, no matter how big the patch is, how much code the patch touches, and how much we want to get it merged to avoid conflicts.

hannes-ucsc commented:
I should have made the risk more clear.
